library/alloc/src/collections/btree/map.rs

```diff
@@ -2437,27 +2435,16 @@ fn handle_underfull_node<'a, K: 'a, V: 'a>(
     }
 }
 
-impl<K: Ord, V, I: Iterator<Item = (K, V)>> Iterator for MergeIter<K, V, I> {
+impl<K: Ord, V, I> Iterator for MergeIter<K, V, I>
+where
+    I: Iterator<Item = (K, V)> + ExactSizeIterator + FusedIterator,
+{
     type Item = (K, V);
 
+    /// If two keys are equal, returns the key/value-pair from the right source.
     fn next(&mut self) -> Option<(K, V)> {
-        let res = match (self.left.peek(), self.right.peek()) {
-            (Some(&(ref left_key, _)), Some(&(ref right_key, _))) => left_key.cmp(right_key),
-            (Some(_), None) => Ordering::Less,
-            (None, Some(_)) => Ordering::Greater,
-            (None, None) => return None,
-        };
-
-        // Check which elements comes first and only advance the corresponding iterator.
-        // If two keys are equal, take the value from `right`.
-        match res {
-            Ordering::Less => self.left.next(),
-            Ordering::Greater => self.right.next(),
-            Ordering::Equal => {
-                self.left.next();
-                self.right.next()
-            }
-        }
```

Comment:
I think this code is simpler than the one we merged with, perhaps we should use it as the canonical one? (In particular using peekable iterators rather than hand-rolling it). Is there a reason you didn't go down that route?

Reply:
We tuned it away from that route in #59186. I had to tune again now because the BTreeSet implementation relied on the copy bound. `Peekable` is relatively costly, even when the items are references like it has for all the set benchmarks, because it has 2 `Option`s.

Reply:
Okay, seems fine. Thanks!